db/system_keyspace.cc

```diff
@@ -1503,6 +1503,27 @@ future<> system_keyspace::cache_truncation_record() {
     });
 }
 
+    future<> system_keyspace::drop_truncation_rp_records() {
+        sstring req = format("SELECT table_uuid, shard, segment_id from system.{}", TRUNCATED);
+        auto rs = co_await execute_cql(req);
+
+        bool any = false;
+        co_await parallel_for_each(rs->begin(), rs->end(), [&] (const cql3::untyped_result_set_row& row) -> future<> {
+            auto table_uuid = table_id(row.get_as<utils::UUID>("table_uuid"));
+            auto shard = row.get_as<int32_t>("shard");
+            auto segment_id = row.get_as<int64_t>("segment_id");
+
+            if (segment_id != 0) {
+                any = true;
+                sstring req = format("UPDATE system.{} SET segment_id = 0, position = 0 WHERE table_uuid = {} AND shard = {}", TRUNCATED, table_uuid, shard);
+                co_await execute_cql(req);
+            }
+        });
+        if (any) {
+            co_await force_blocking_flush(TRUNCATED);
+        }
+    }
+
```

Comment:
Is this really needed? Sure the replay positions are unnecessary, but are they really invalid? Assuming we switch to a better clock.

(though, it's impossible to guarantee the wall clock is monotonic)

Reply:
[Here is ](https://github.com/scylladb/scylladb/pull/15532#issuecomment-1735636564) a scenario when they become invalid with current `steady_clock`. I'm not sure switching the clock would be simpler. What's the problem with removing (invalidating) truncation records?

Reply:
Just fear of the unknown. I now think removing the truncation records is the best option.